debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/RowDeserializers.java

```diff
@@ -517,12 +538,56 @@ protected static int bitSlice(long value, int bitOffset, int numberOfBits, int p
      */
     protected static int deserializeFractionalSecondsInNanos(int fsp, ByteArrayInputStream inputStream) throws IOException {
         // Calculate the number of bytes to read, which is
+        // '1' when fsp=(1,2) -- 7
+        // '2' when fsp=(3,4) and -- 12
+        // '3' when fsp=(5,6)  -- 21
+        int length = (fsp + 1) / 2;
+        if (length > 0) {
+            long fraction = bigEndianLong(inputStream.read(length), 0, length);
+            // Convert the fractional value (which has extra trailing digit for fsp=1,3, and 5) to nanoseconds ...
+            return (int) (fraction / (0.0000001 * Math.pow(100, length - 1)));
+        }
+        return 0;
+    }
+
+    /**
+     * Read the binary input stream to obtain the number of nanoseconds given the <em>fractional seconds precision</em>, or
+     * <em>fsp</em>.
+     * <p>
+     * We can't use/access the {@code deserializeFractionalSeconds} method in the {@link AbstractRowsEventDataDeserializer} class,
+     * so we replicate it here with modifications to support nanoseconds rather than microseconds and negative values.
+     * Note the original is licensed under the same Apache Software License 2.0 as Debezium.
+     *
+     * @param fsp the fractional seconds precision describing the number of digits precision used to store the fractional seconds
+     *            (e.g., 1 for storing tenths of a second, 2 for storing hundredths, 3 for storing milliseconds, etc.)
+     * @param inputStream the binary data stream
+     * @return the number of nanoseconds
+     * @throws IOException if there is an error reading from the binlog event data
+     */
+    protected static int deserializeFractionalSecondsInNanosNegative(int fsp, ByteArrayInputStream inputStream) throws IOException {
```

Comment:
Why is it we need to deal with nano-seconds at all? IIUC, MySQL supports only micro-second resolution for `TIME`.

Reply:
see my first reply.

Reply:
@gunnarmorling btw, the recent code was calculating nanoseconds here too.